Advangelists Prebid Server Bidder

trchandraprakash opened the pull request.

trchandraprakash commented:
Hi @mansinahar, Let me know if any changes are needed and when we can make it available for download. Thanks

trchandraprakash commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,269 @@
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const defaultDomain string = "nep.advangelists.com"
```
Done.

trchandraprakash commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,269 @@
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
+	if len(dispErrors) > 0 {
```
Agree. Done

trchandraprakash commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,269 @@
+	return openrtb_ext.BidTypeBanner
+}
+
+func newBadInputError(message string) error {
```
Have update code accordingly. Done

trchandraprakash commented on config/config.go:
```diff
@@ -185,7 +185,9 @@ func (cfg *HostCookie) TTLDuration() time.Duration {
 	dummyPublisherID int    = 12
+	dummyPubid       string = "dummypubid"
 	dummyGDPR        string = "0"
+
```
Done

trchandraprakash commented on macros/macros.go:
```diff
@@ -8,6 +8,7 @@ import (
 // EndpointTemplateParams specifies params for an endpoint template
 type EndpointTemplateParams struct {
 	Host        string
+	Pubid       string
```
Pubid is string.

mansinahar replied:
Okay. Is it just a string of numbers or it could be a string of characters or mix of characters and numbers too? Even if that's the case, it might still be easier to have just `PublisherID` and make that a string rather than have `PubID` and `PublisherID` both. We can easily change an integer publisher ID to a string in that case.

mansinahar replied:
Ah actually I see from your example json that `PubID` could be a mix of characters and numbers. In that case as suggested above, could you please rather change `PublisherID` to type string itself and the integer `PublisherID` values used before can be converted into string as needed?

mansinahar replied:
@trchandraprakash pinging to see if you have had a chance to make the above change

trchandraprakash replied:
@mansinahar I am OOO this week. Will update coming monday. Also if I change PublisherID macro to be a string, won't it impact other bidder adapters who are considering it as integer.

mansinahar replied:
Well, yeah those values will have to be converted into string before setting them to `PublisherID` and this PR would have to do that work. From a quick look hopefully it shouldn't be a lot of changes though

trchandraprakash commented on openrtb_ext/imp_advangelists.go:
```diff
@@ -0,0 +1,5 @@
+package openrtb_ext
+
+type ExtImpAdvangelists struct {
+	Pubid string `json:"pubid"`
```
Done

trchandraprakash commented:
Hi Mansi, Can you please tell what causing checks fail. Same command works on my environment.
./validate.sh --nofmt --cov --race 10

mansinahar commented:
Hi @trchandraprakash! Looks like there are formatting issues. I'd suggest running `go fmt` on your branch and then updating the PR. Specifically the build is complaining about `openrtb_ext/imp_advangelists.go`. That should hopefully resolve the issue.

trchandraprakash commented:
Got it thanks. Let me know if it looks good.

guscarreon commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,252 @@
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
+	//As banner.w/h are required fields for advangelistsAdn platform - take the first format entry
+	if banner.W == nil && banner.H == nil {
```
Are BOTH fields required? In other words, if the truth table below is accurate, shouldn't we switch the boolean operator to a short circuit OR ( || )?
```
banner.W |  banner.H |   if statement
-----------------------------------------
  nil    |    nil    | into true branch 
  nil    |  not nil  | into true branch 
 not nil |    nil    | into true branch 
 not nil |  not nil  |    skip          
```

trchandraprakash replied:
agree, updated. thanks

guscarreon commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,252 @@
+	}
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
```
We should append `dispErrors` before returning `errs` because it will probably be populated if `len(pub2impressions) == 0`. In other words `errs = append(errs, dispErrors...)` should be located in line 35 and not 38.

trchandraprakash replied:
done.

guscarreon commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,252 @@
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
```
Are we sure we want to return a non-NULL `result` array even if there are some errors in `errs`?

trchandraprakash replied:
made required change. thanks.

trchandraprakash replied:
Hi, Let me know if anything else need to be taken care, I have make it available asap. thanks.

trchandraprakash commented:
Hi, Please let me know if it is all good, I have to make it available asap. thanks.

trchandraprakash commented:
Hi @guscarreon, I have made requested changes, can you please help in finalizing review, so that i can make it available to our clients asap.

trchandraprakash commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,252 @@
+	res := make(map[openrtb_ext.ExtImpAdvangelists][]openrtb.Imp)
+	errors := make([]error, 0)
+	for idx := range imps {
+		imp := imps[idx]
```
done as suggested.

trchandraprakash commented on adapters/advangelists/advangelists.go:
```diff
@@ -0,0 +1,252 @@
+	impsCount := len(imps)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	resImpExts := make([]openrtb_ext.ExtImpAdvangelists, 0, impsCount)
```
done as requested.

mansinahar commented:
@trchandraprakash Thanks for addressing all the feedback. If you could please make the `PubID` change when you have a chance, we can definitely help merge and close this ASAP.

trchandraprakash commented:
Hi @mansinahar, was not sure if i should modify other bidder code, but i guess without that i cannot pass automated check. Have done some modifications here
1. Updated macro PublisherID to be string.
2. Updated my bidder "Advangelists" code to use PublisherID macro instead of PubID.
3. Updated another bidder "adkernelAdn" to accept PublisherID as string.
Let me know if this is good. Thanks.

mansinahar commented:
@trchandraprakash I think this PR should just convert the `PublisherID` in the "AdKernel" adapter right before the `ResolveMacros` function is called here: https://github.com/prebid/prebid-server/blob/master/adapters/adkernelAdn/adkernelAdn.go#L203. Something like: 

```
pubIDStr := strconv.Itoa(params.PublisherID)
endpointParams := macros.EndpointTemplateParams{Host: reqHost, PublisherID: pubIDStr}
return macros.ResolveMacros(adapter.EndpointTemplate, endpointParams)
```

The reason being, the change that you've made is fine but is sort of a big change for their adapter and we would have to get their approval also before merging it in. That will just delay this more for you so we can rather do this simple workaround for now and they can submit another PR changing their adapter to accept a string `PublisherID`

trchandraprakash commented:
@mansinahar I see what you mean, i have made requested changes and reverted back others.

trchandraprakash commented:
@mansinahar when we can see it available on production for client. Also not sure why i still see status of this PR as changes requested.

mansinahar commented:
@trchandraprakash I think those changes requested are from Gus's review. We can merge and close once we get a second approval from @guscarreon too.

trchandraprakash commented:
Thanks @mansinahar @guscarreon for approval. Can you please suggest how or when adapter will be available on this page http://prebid.org/dev-docs/prebid-server-bidders.html.

mansinahar commented:
@trchandraprakash We are looking to cut a new release on our end tomorrow so it should be available in prod thereafter. We will keep you updated. As for it appearing on http://prebid.org/dev-docs/prebid-server-bidders.html, I think @bretg might be responsible for that or at least should be able to point to the person who is.

trchandraprakash commented:
@mansinahar thanks will wait for your update. @bretg  can you please help on getting adapter on http://prebid.org/dev-docs/prebid-server-bidders.html.

mansinahar commented:
@trchandraprakash FYI: We cut a new tag - [0.76.0](https://github.com/prebid/prebid-server/releases/tag/0.76.0) yesterday but unfortunately were having some issues with the release and had to rollback. We're planning to roll it out again on Monday and will keep you posted.

trchandraprakash commented:
@mansinahar thanks for update.

bretg commented:
For the record, the list of adapters on http://prebid.org/dev-docs/prebid-server-bidders.html is dynamically generated from adapters that are deployed to AppNexus' cluster of servers. So when that deploy succeeds, the adapter will be on the list.

trchandraprakash commented:
Thanks @bretg . 
Hi @mansinahar, Any updates, can we expect adapter to be available today?

mansinahar commented:
@trchandraprakash The adapter is now available in production and on the list of adapters on http://prebid.org/dev-docs/prebid-server-bidders.html

trchandraprakash commented:
@mansinahar thanks a lot.
